### PR TITLE
feat: [IDP-804]: Users and UserGroups sync optimization

### PR DESCRIPTION
**Pull request by sathish-soundarapandian**
## Describe your changes
Users and UserGroups sync optimization

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51025)

### CODE REVIEW

**Review by vikyathharekal — APPROVED** [review]

**Review by vikyathharekal — DISMISSED** [review]

File: idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+          updateOperationResult.getModifiedCount());
+    } else {
+      log.warn("Could not add userGroupIdentifier field and set hasEvent field to false for any record");
+    }
+
+    try {
+      mongoPersistence.getCollection(UserEventEntity.class).dropIndex("accountIdentifier_1");
+      log.info("Dropped accountIdentifier_1 index on idp-harness.userEvents collection");
+    } catch (Exception ex) {
+      log.info("Error dropping accountIdentifier_1 index on idp-harness.userEvents collection. Error = {}",
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java:50]: nit: If dropIndex() fails, should we log a warning or error instead?

**Review by vikyathharekal — APPROVED** [review]

File: idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java (hunk 1 above)
Comment (sathish-soundarapandian) [reply idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java]: Done

**Review by vikyathharekal — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger ut0

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger ut4, ut5

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 7 of 12 files at r1, 1 of 1 files at r3, 1 of 2 files at r5, 3 of 3 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51025)* status: 13 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java` line 33 at r8](https://reviewable.io/reviews/harness/harness-core/51025#-NcR4-ew7aak-J6fnsAt:-NcR4-exDVHVj1H6e5yl:b-91jmot) ([raw file](https://github.com/harness/harness-core/blob/d55d6ab694e5674bd5ef85c3f7f0afc9f1f965a4/idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java#L33)):*
> ```Java
> 
>     BasicDBObject basicDBObject = new BasicDBObject();
>     BasicDBObject updateOps = new BasicDBObject(UserEventEntity.UserEventKeys.userGroupIdentifier, "")
> ```

If there's an entry with hasEvent as true currently in the DB, then we would lose that and not trigger a user sync right?

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51025)* status: 13 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java` line 33 at r8](https://reviewable.io/reviews/harness/harness-core/51025#-NcR4-ew7aak-J6fnsAt:-NcR5nxA1ZQh-KSjWNn_:b-z37icn) ([raw file](https://github.com/harness/harness-core/blob/d55d6ab694e5674bd5ef85c3f7f0afc9f1f965a4/idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java#L33)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

If there's an entry with hasEvent as true currently in the DB, then we would lose that and not trigger a user sync right?
</blockquote></details>

Yes. During deployment, we can check it once. Anyways, in the next 15 mins, the full trigger will make up for it.

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51025)* status: 13 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java` line 33 at r8](https://reviewable.io/reviews/harness/harness-core/51025#-NcR4-ew7aak-J6fnsAt:-NcR6iVqBudvWzikkXGR:b-k3dwfb) ([raw file](https://github.com/harness/harness-core/blob/d55d6ab694e5674bd5ef85c3f7f0afc9f1f965a4/idp-service/src/main/java/io/harness/idp/migration/UserEventEntityUserGroupIdentifierMigration.java#L33)):*
<details><summary><i>Previously, sathish-soundarapandian (Sathish) wrote…</i></summary><blockquote>

Yes. During deployment, we can check it once. Anyways, in the next 15 mins, the full trigger will make up for it.
</blockquote></details>

Moreover, we will deploy idp-service first and then harness-idp. So during the startup, harness-idp will do a full sync.

**Review by vikyathharekal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51025)* status: 13 file(s) to review. 1 of 2 approvals obtained.

**Comment by devesh-harness** [comment]
trigger review

**Pull request merged**